secure-api-gateway-ig-extensions/src/main/java/com/forgerock/sapi/gateway/dcr/request/DCRSsaSignatureValidator.java

```diff
@@ -0,0 +1,155 @@
+     */
+    public Promise<Response, DCRSignatureValidationException> validateSoftwareStatementAssertionSignature(
+            @NotNull String transactionId,
+            @NotNull SignedJwt ssaSignedJwt) {
+        try {
+
+            URL issuingDirectoryJwksUrl = getIssuingDirectoryJwksUrl(transactionId, ssaSignedJwt);
+            return this.jwkSetService.getJwkSet(issuingDirectoryJwksUrl).thenAsync(directoryJwkSet -> {
+                try {
+                    this.jwtSignatureValidator.validateSignature(ssaSignedJwt, directoryJwkSet);
+                    log.debug("({}) SSA has a valid signature", transactionId);
+                    return Promises.newResultPromise(new Response(Status.OK));
+                } catch (SignatureException e) {
+                    String errorDescription = "Failed to validate SSA against jwks_uri '" + issuingDirectoryJwksUrl +
+                            "'";
+                    log.debug("({}) {}", transactionId, errorDescription);
+                    return Promises.newExceptionPromise(
+                        new DCRSignatureValidationException(ErrorCode.INVALID_SOFTWARE_STATEMENT, errorDescription));
+                }
+            }, ex -> {
+                String errorDescription = "Failed to obtain jwk set from trusted directory uri " +
+                        issuingDirectoryJwksUrl;
+                log.debug("({}) {}}", transactionId, errorDescription, ex);
+                return Promises.newExceptionPromise(
+                        new DCRSignatureValidationException(ErrorCode.INVALID_SOFTWARE_STATEMENT, errorDescription));
+            });
+        } catch (DCRSignatureValidationException dcre) {
+            return Promises.newExceptionPromise(dcre);
+        }
+
+    }
+
+    /**
+     * get the URL at which the JWKS for the SSA can be found
+     * @param transactionId log entries will contain the transactionId
+     * @param ssa the Software Statement Assertion from which to obtain the JWKS URL
+     * @return the URL at which the Software Statement Assertion's signature may be validated
+     * @throws DCRSignatureValidationException when the URL can't be obtained from the Software Statement
+     */
+    private URL getIssuingDirectoryJwksUrl(String transactionId, SignedJwt ssa) throws DCRSignatureValidationException {
+        JwtClaimsSet ssaClaimSet = ssa.getClaimsSet();
+
+        final String JWT_NAME = "software statement assertion";
+        String ssaIssuer = dcrUtils.getJwtIssuer(JWT_NAME, ssa);
+        log.debug("({}) {} issuer is {}", transactionId, JWT_NAME, ssaIssuer);
+
+        TrustedDirectory ssaIssuingDirectory = dcrUtils.getIssuingDirectory(trustedDirectoryService, ssaIssuer);
+
+        String issuingDirectoryJwksUriString = ssaIssuingDirectory.getDirectoryJwksUri();
+        return issuingDirectoryJwksUriFromString(transactionId, issuingDirectoryJwksUriString, ssaIssuer);
```

Comment:
I think it makes sense to update the return type of com.forgerock.sapi.gateway.trusteddirectories.TrustedDirectory#getDirectoryJwksUri to URL, then we dodgy Trusted Directory config will be rejected during the configuration phase and we do not need to handle the case on the runtime hot path